extensions/resume.md

```diff
@@ -0,0 +1,163 @@
+
+### Messages
+
+There are two new messages defined by this specification:
+
+- `RESUME`: Indicates that the client wishes to resume a previous connection.
+- `RESUMED`: Indicates that a client has disconnected+reconnected to the server.
+
+Below, we go through both of these.
+
+#### `RESUME` Command
+
+This command is sent before starting SASL authentication, and indicates that the client wishes to resume a previous connection to the server which may still be active. The command is sent with the following format:
+
+    RESUME oldnick [timestamp]
+
+`oldnick` is the nickname the client had when they last disconnected. `timestamp`, if it is given, is a timestamp which indicates when they received the last message from the server on the old connection. This timestamp is used to indicate to other clients how long the disconnection was for, and uses the same format as the IRCv3 `server-time` extension (i.e. `YYYY-MM-DDThh:mm:ss.sssZ`, or in UTC using extended format as specified by ISO 8601:2004(E) 4.3.2).
+
+After sending this command, the client continues connection registration as usual.
+
+If the client's able to successfully resume their connection, the server sends them a `RESUMED` message before sending the usual registration burst (and their nickname is set to `<oldnick>`). Otherwise, the server sends the `ERR_CANNOT_RESUME` numeric (and optional additional error numerics).
+
+If successful, the `RESUMED` message is sent to the connecting user just after connection registration (including ident lookups and similar) completes. When this message is sent out, the connecting client's nickname is updated to be the old client's nick. From this point forward, the combination of old nickname and new connecting client's username+hostname is used to identify the client from that point forward.
+
+Connection resumption is done on a best-effort basis. If the resumption fails clients SHOULD continue with a regular reconnection.
+
+#### `RESUMED` Message
+
+This message is sent when a client successfully resumes a connection. For clients that have not received numerics indicating that their connection registration is completed, this numeric means that their `RESUME` attempt was successful. For registered clients, this indicates that the given client disconnected and reconnected to the server.
+
+This message has the following format:
+
+    :nick!olduser@oldhost RESUMED nick user host [timestamp]
+
+`nick` is the nickname of the client whose connection has been resumed. `user` and `host` are the username and hostname that the new connecting client has. The client receiving this message MUST update the username and hostname they have stored for the given client, in the same way that the `CHGHOST` message is parsed. `timestamp` if given, shows when the reconnecting client received their last message from the server. The timestamp may be used as a rough indication of how much message history has been lost.
```

Comment:
To better support `chathistory`, allowing either timestamps or `draft/msgid` to be sent could be useful, as `chathistory` allows clients to request content by either parameter.

    :nick!olduser@oldhost RESUMED nick user host [timestamp=2017-04-13T15:12:51.620Z]
    :nick!olduser@oldhost RESUMED nick user host [draft/msgid=123456789]

Reply:
The `[timestamp]` value only allows other clients to know how much message history may have been lost. The reconnecting client will attempt to retrieve their missing history themselves (and/or get it replayed), they don't use or depend on this value in any way so allowing them to send a message ID in this space would only complicate things for those other clients.